builder/dockerfile/command/commands.go

```diff
@@ -0,0 +1,232 @@
+package command
+
+import (
+	"errors"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/strslice"
+)
+
+type KeyValuePair struct {
+	Key   string
+	Value string
+}
+
+type KeyValuePairs []KeyValuePair
+
+// WithSourceCode is a marker indicating that a given command
+// Has been parsed from source code (wich can be displayed in builder output)
+type WithSourceCode interface {
+	SourceCode() string
+}
+
+type EnvCommand struct {
+	OriginalSource string
+	Env            KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+func (c *EnvCommand) SourceCode() string {
+	return c.OriginalSource
+}
+
+type MaintainerCommand struct {
+	OriginalSource string
+	Maintainer     string
+}
+
+func (c *MaintainerCommand) SourceCode() string {
+	return c.OriginalSource
+}
+
+type LabelCommand struct {
+	OriginalSource string
+	Labels         KeyValuePairs // kvp slice instead of map to preserve ordering
+}
+
+func (c *LabelCommand) SourceCode() string {
+	return c.OriginalSource
+}
+
+type AddCommand struct {
+	OriginalSource string
+	Srcs           []string
+	Dest           string
+}
+
+func (c *AddCommand) SourceCode() string {
+	return c.OriginalSource
+}
+
+type CopyCommand struct {
+	OriginalSource string
+	Srcs           []string
+	Dest           string
+	From           string
+}
+
+func (c *CopyCommand) SourceCode() string {
+	return c.OriginalSource
+}
+
+type FromCommand struct {
+	OriginalSource string
+	BaseName       string
+	StageName      string
```

Comment:
I introduced it because It feels more flexible on the dispatch side to have the first command disptacher initiate the dispatch operation (initialization case with BuildFromConfig seemed a bit confusing to me, and in anticipation of build parallelism, it felt better to have `from` be a dispatched command as it might trigger a wait for an other stage to complete)